### PR TITLE
Add Skip to Content to Application Header Layout

### LINKED ISSUE #266: Add Skip to Content button for Application Layout
# Feature Request

## Description
To improve accessibility of our UI for users using a keyboard to navigate, we should implement a skip to content button, similar to what airbnb does.

## Additional Context / Screenshots
screen shot 2018-09-14 at 10 17 48 am

One thing to note is that we will need to attach an ID element to both the header navigation and the main node inside the layout. Should we implement this a a prop that can be overridden, and default the prop to be an id of main?

### PR DESCRIPTION
**Pull request by JakeLaCombe**
### Summary
Fixes #266

To make Terra more accessible, we need to have a skip to content button that automatically focuses on the main content node. To use it, just tab onto the page, and you'll see it as the first item.

### Additional Details
Right now, the scroll is not working. However, this has to do less with this button, and more with the scroll container. I'll be investigating how we can better handle the scroll container. As of right now, to test the behavior, activate the skip to content button, then hit tab to get to the next focusable area of the container.

### CODE REVIEW

File: packages/terra-application-header-layout/tests/jest/ApplicationHeader.test.jsx [hunk 1]
```diff
@@ -1,121 +1,175 @@
 import React from 'react';
+import { IntlProvider } from 'react-intl';
```
Comment (bjankord) [inline packages/terra-application-header-layout/tests/jest/ApplicationHeader.test.jsx]: You can import this from terra-base, and then you can remove the devDependency on react-intl.

**Comment by JakeLaCombe** [comment]
That was my fault on it. My latest commit should have that error fixed.

File: packages/terra-application-header-layout/package.json [hunk 2]
```diff
@@ -31,6 +31,7 @@
   "dependencies": {
     "classnames": "^2.2.5",
     "prop-types": "^15.5.8",
+    "terra-base": "3.24.0",
```
Comment (bjankord) [inline packages/terra-application-header-layout/package.json]: Let's add a caret before this version. `^3.24.0`

File: packages/terra-application-header-layout/translations/sv.json [hunk 3]
```diff
@@ -0,0 +1,3 @@
+{
+  "Terra.ApplicationHeaderLayout.SkipToContent": "Gå till Innehåll"
+}
```
Comment (emilyrohrbough) [inline packages/terra-application-header-layout/translations/sv.json]: We need translations for `sv-SE` translations as well

File: packages/terra-application-header-layout/translations/pt.json [hunk 4]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (emilyrohrbough) [inline packages/terra-application-header-layout/translations/pt.json]: We need `pt-BR` translations as well

File: packages/terra-application-header-layout/translations/nl.json [hunk 5]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (emilyrohrbough) [inline packages/terra-application-header-layout/translations/nl.json]: We need `nl-BE` translations as well

File: packages/terra-application-header-layout/translations/fr.json [hunk 6]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (emilyrohrbough) [inline packages/terra-application-header-layout/translations/fr.json]: We need `fr-FR ` translations as well

File: packages/terra-application-header-layout/translations/es.json [hunk 7]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (emilyrohrbough) [inline packages/terra-application-header-layout/translations/es.json]: We need `es-US` and `es-ES` translations as well

File: packages/terra-application-header-layout/translations/de.json [hunk 8]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (emilyrohrbough) [inline packages/terra-application-header-layout/translations/de.json]: We need `ar` translatios as well

File: packages/terra-application-header-layout/translations/sv.json (hunk 3 above)
Comment (bjankord) [reply packages/terra-application-header-layout/translations/sv.json]: https://github.com/cerner/terra-framework/pull/301/commits/5c99136fb9295b0edf9e4b6da11ad8d01d4c7b57

File: packages/terra-application-header-layout/translations/pt.json (hunk 4 above)
Comment (bjankord) [reply packages/terra-application-header-layout/translations/pt.json]: https://github.com/cerner/terra-framework/pull/301/commits/5c99136fb9295b0edf9e4b6da11ad8d01d4c7b57

File: packages/terra-application-header-layout/translations/nl.json (hunk 5 above)
Comment (bjankord) [reply packages/terra-application-header-layout/translations/nl.json]: https://github.com/cerner/terra-framework/pull/301/commits/5c99136fb9295b0edf9e4b6da11ad8d01d4c7b57

File: packages/terra-application-header-layout/translations/fr.json (hunk 6 above)
Comment (bjankord) [reply packages/terra-application-header-layout/translations/fr.json]: https://github.com/cerner/terra-framework/pull/301/commits/5c99136fb9295b0edf9e4b6da11ad8d01d4c7b57

File: packages/terra-application-header-layout/translations/es.json (hunk 7 above)
Comment (bjankord) [reply packages/terra-application-header-layout/translations/es.json]: https://github.com/cerner/terra-framework/pull/301/commits/5c99136fb9295b0edf9e4b6da11ad8d01d4c7b57

File: packages/terra-application-header-layout/translations/de.json (hunk 8 above)
Comment (bjankord) [reply packages/terra-application-header-layout/translations/de.json]: https://github.com/cerner/terra-framework/pull/301/commits/932608d8f23365316a8f4bf72b180e68907596c3

File: packages/terra-application-header-layout/package.json [hunk 9]
```diff
@@ -31,6 +31,7 @@
   "dependencies": {
     "classnames": "^2.2.5",
     "prop-types": "^15.5.8",
+    "terra-base": "^3.24.0",
```
[the remark this answers is not in the document]
Comment (JakeLaCombe) [reply packages/terra-application-header-layout/package.json]: We do use the intlShape and intlProvider from the terra-base, and without including it here it looks like the package can not find it.

File: packages/terra-application-header-layout/src/ApplicationHeaderLayout.jsx [hunk 10]
```diff
@@ -79,10 +85,21 @@ const ApplicationHeaderLayout = ({
     );
   }
 
+  const skipToContent = () => {
+    document.getElementById('terra-main').focus();
+  };
+
+  const skipToContentButton = (
+    <button type="button" onClick={skipToContent} className={cx('skip-content')}>
```
Comment (emilyrohrbough) [inline packages/terra-application-header-layout/src/ApplicationHeaderLayout.jsx]: Should we be using Terra-button for this update?

File: packages/terra-application-header-layout/tests/jest/ApplicationHeader.test.jsx [hunk 11]
```diff
@@ -1,121 +1,175 @@
 import React from 'react';
+import { IntlProvider } from 'terra-base';
 
 import ApplicationHeaderLayout from '../../src/ApplicationHeaderLayout';
 
 describe('ApplicationHeaderLayout', () => {
   it('should render without optional props', () => {
-    const result = shallow(<ApplicationHeaderLayout />);
+    const result = shallow(
+      <IntlProvider locale="en">
```
Comment (emilyrohrbough) [inline packages/terra-application-header-layout/tests/jest/ApplicationHeader.test.jsx]: We should be able to pass the intl context when rendering this component so we don't have to wrap it in the IntlProvider. Here is an example for reference:
- Setup Intl context: https://github.com/cerner/terra-core/blob/master/packages/terra-tabs/tests/jest/intl-context-setup.js
- Use in Jest test: https://github.com/cerner/terra-core/blob/master/packages/terra-tabs/tests/jest/Tabs.test.jsx#L8

File: packages/terra-layout/src/_LayoutSlidePanel.jsx [hunk 12]
```diff
@@ -151,7 +151,7 @@ class LayoutSlidePanel extends React.Component {
         </div>
         <OverlayContainer className={cx('content')}>
           <Overlay isRelativeToContainer onRequestClose={onToggle} isOpen={isOverlayOpen} backgroundStyle={overlayBackground} />
-          <main className={cx('main-container')}>
+          <main id="terra-main" tabIndex="-1" className={cx('main-container')}>
```
Comment (emilyrohrbough) [inline packages/terra-layout/src/_LayoutSlidePanel.jsx]: Should we be providing a data-attribute for this instead? Like `data-terra-layout-main`? In terra-table we pass [`data-terra-table-cell`](https://github.com/cerner/terra-core/blob/master/packages/terra-table/src/TableCell.jsx#L17)

File: packages/terra-application-header-layout/src/ApplicationHeaderLayout.jsx (hunk 10 above)
Comment (JakeLaCombe) [reply packages/terra-application-header-layout/src/ApplicationHeaderLayout.jsx]: Brett and I talked. Given that this button is visually hidden most of the time, it doesn't make sense to inject a dependency requirement just for this one button.

File: packages/terra-application-header-layout/tests/jest/ApplicationHeader.test.jsx (hunk 11 above)
Comment (JakeLaCombe) [reply packages/terra-application-header-layout/tests/jest/ApplicationHeader.test.jsx]: We actually aren't using the context for intl anymore, so I don't think this approach would help us. We did agree to using the IntlProvider on this code review. https://github.com/cerner/terra-core/pull/1956/files

File: packages/terra-layout/src/_LayoutSlidePanel.jsx (hunk 12 above)
Comment (JakeLaCombe) [reply packages/terra-layout/src/_LayoutSlidePanel.jsx]: Got it changed. I like that a lot better than having the id on the main tag!

**Pull request merged**